### PR TITLE
Http1.1: Retain keep alive for response headers and connection keep-alive for request headers

### PR DESCRIPTION
**Pull request by noah8713**
Rely on keep-alive timeout settings from upstream server.
If upstream emits keep-alive while using http1.1, its being removed by envoy. 
Hence, retain
`keep-alive: timeout=x` in response header..

This will help reduce load on envoy cpu as ssl connections can be reused vs creating new connections for each calls.
If  http protocol for gw<->sidecar is  h2 in case of mesh, this will not take effect even if runtime guard is true.

upstream issue https://github.com/envoyproxy/envoy/issues/17613

1. Fix existing UTs
2. Also add unit test to retain keep alive for http1.1 response header + connection keep alive in request header
3. Part of tests client using http1.1 for java apps using tomcat, client relies on `connection: keep-alive `request header. Hence, retain the same as well for http1.1 for upstream to add keep-alive back in response header
4. Enable this via HCM config  `retain_keepalive_header:true`. Setting it to false will revert to current behavior

### CODE REVIEW

**Comment by noah8713** [comment]
Tests with nodejs server
```
curl --http1.1 -kv https://foo-nodejs.com:443 | grep -i keep-al
Xxxx
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
Xxx
< keep-alive: timeout=5
```
Tests with java/tomcat as it needs connection:kee-alive
```
curl --http1.1 -kv https://foo-java.com:443 -H "Connection: keep-alive" | grep -i keep-al
< HTTP/1.1 404 Not Found
xxxx
< keep-alive: timeout=20
```

**Comment by noah8713** [comment]
cc ptal @alyssawilk

**Comment by noah8713** [comment]
Also cc @envoyproxy/runtime-guard-changes and @phlax @lizan PTAL/ping

File: source/common/http/conn_manager_utility.cc
Comment (alyssawilk) [inline source/common/http/conn_manager_utility.cc:96]: I don't think this is the route we want to go. Upstream keep-alive and keep-alive duration should be seprate from downstream keep-alive.  I think it'd be fine to add a keep alive to the response sent downstream but I'm unsure why we'd want to proxy one through.  Can you be a bit more clear about why you want the same timeout persisted?
If you go this route it 100% has to be a long term config option and be added to HCM API surface, as I don't think most folks will want it so reloadble flags don't seem optimal.

**Review by alyssawilk — COMMENTED** [review]
Hey thanks for picking this up.  one early question, and will wait on your response
/wait

File: source/common/http/conn_manager_utility.cc
Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Thanks  for review.  I think sending keep-alive from upstream response header is neat. Issue is different servers have different behavior e.g. if upstream is java/tomcat based, it will not emit keep-alive timeout settings in response  header by default until client sends `connection:keep-alive` which jdk clients does  but it gets dropped by proxy. We are not honoring downstream keep-alive time out in request header, just connection header. PTAL at ref test doc for nodejs and java/tomcat behind proxy https://docs.google.com/document/d/1JNynMtGAy5czphdJzvutgoVKG3UGqfo3CMVheea6O-E/edit . I am ok to skip `connection:keep-alive `in request header if can call out that customers using apache/tomcat behind proxy/envy, tweaksdirectly on the server side to send keep-alive response header by default similar to how nodejs does. As long as upstream emits, envoy can honor that part of this change. 

May be long term sure can explore to convert to HCP API vs runtime guard for now.

Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc:96]: I'm still unclear why we want to use the same keep-alive.  The issue you link (https://github.com/envoyproxy/envoy/issues/17613) involves Envoy setting its own keep-alive header to make it clear to clients when Envoy will close the connection, which seems reasonable.  Envoy forwarding an upstream keep-alive timeout which doesn't correspond to any Envoy timeout configuration continues to make no sense to me.   The network set up here is upstream ---- Envoy ---- client.  The upstream to Envoy connection has a completely different lifetime than the Envoy to client connection, and unless you can give a compelling reason to forward one connection's timeout to another unrelated connection, I don't think this change should proceed.  If you do have a compelling reason, it is not OK for this to land as a runtime guarded change, as it violates Envoy data plane principles and would likely cause harm to users.  It needs to be a config guarded option if it lands at all.

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Thanks.Seems with hardware load balancers, it forwards the same server keep-alive as is until we explicitly over-ride .Hence, kept the behavior intact vs we can also add some default keep-alive capability based on say idletimeout settings, etc. but this seems more neat to align what backend/upstream has in its default. Because after move to proxy based LB, we see ssl connection creation spikes up as keep-alive timeout itself is missing causing clients to create new ssl connections for each requests vs reuse causing more CPU load on envoy. So I am ok to change timeout values later too but original intent to pass the in response header still stays same. Let me know further 

So do you want me to make this as a config guard vs runtime in this patch or an incremental change is ok too? Please let me know.

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: @alyssawilk hello just following up on this again. I dialed in the community meeting and seems there wasnt any audience so wanted to have a gentle reminder as we use this patch internally and helps overall. Ok to do it incrementally by returning upstream default keep-alive for h1.1 and have an add on to send envoy's default. Reason is returning just envoy's default will not help if upstream/server is not honoring it for e2e flow. Let me know cc @mattklein123

Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc:96]: Hardware load balancers may actually tie the lifetime of the two connections, Envoy does not so that argument still isn't particularly compelling.  Meanwhile this PR is marked waiting and will not get further review until you addressed the comments in my last pass, which you have not done.

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Thanks a bunch. I was waiting to clarify. So we we ok with this patch but address it with config changes vs runtime guard? Please advice and I will rectify accordingly to make it config guard change vs runtime guard as we use it internally via runtime guard and works fine too. can update the ssl metrics and cpu improvements results in the same doc. Let me know.

Comment (kb000) [reply source/common/http/conn_manager_utility.cc:96]: @noah8713 It looks like we should add something like `connection_headers_to_preserve: ['keep-alive']` to [HttpConnectionManager](https://github.com/envoyproxy/envoy/blob/152cf3d30319fc739cd84e47f3147eec6d1b54b4/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L41) proto.  Though this behavior would violate the connection-token removal requirements of https://datatracker.ietf.org/doc/html/rfc2616#section-8.1.3, we have definitely seen material improvements in the controlled environment of our service mesh.

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Thanks @alyssawilk for confirming. @kb000 I think it would be good to expose similar to case preserve use case via HCM under http_protocol_options where we have [idle timeouts and other bits present](https://github.com/envoyproxy/envoy/blob/main/api/envoy/api/v2/core/protocol.proto#L72). Can move retain_keepalive_header_http11 option under the same and see how it goes.

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Another thing I completely missed is runtime guard is supported as an option from istio via [runtimevalues](https://github.com/istio/istio/blob/master/pkg/bootstrap/config.go#L320) so reverting the HCM config will incur further changes to honor this either via envoyfilter CRD or via some additional changes in istio. I see there are too many guards for http bits e.g. skip_delay_close, etc. So can we keep it if so? I should have called out earlier. Let me know if we can keep runtime guard for first cut as its not a major change too and works fine in our service mesh with no additional changes in istio side. If its a big no, I will proceed to make HCM side change. Please confirm,

Comment (noah8713) [reply source/common/http/conn_manager_utility.cc:96]: Thanks for suggestions. I moved this as a config under HCM. PTAL @alyssawilk

Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc:96]: As said I'm going to defer to the Envoy API shephard, @mattklein123

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto
Comment (mattklein123) [inline api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:525]: Assuming we move forward with this change, this should be reverted. The v2 APIs are dead.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: I think this needs to be a much longer explanation. Specifically, it should clear say which RFC this violates, why it violates it, and when someone would want to set this.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Given this helps real customer use cases and I see multiple people asking for it, it's opt in, and the change surface is small, I'm OK with this.

As suggested in one of the comment threads, it seems like another way of solving this would be to allow setting specific connection headers on different halves of the connection (I think this has been asked for elsewhere), but that would require more manual configuration than this.

/wait

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: Also, this change keeps the request portion of the change also, so if we are keeping that, this name should change also, probably to `retain_keepalive_headers`. The longer comment should also clearly indicate what happens if either half is HTTP/2.

Comment (alyssawilk) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: If setting specific connection headers would be sufficient for this user's use case I'd  much rather go that route.  @noah8713 would set keep-alive work for you or do you actually need preservation of different headers from different upstreams?

Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: See also https://github.com/envoyproxy/envoy/issues/3011 and https://github.com/envoyproxy/envoy/issues/17613.

I would be in favor of giving users control of connection headers if they wish per these linked issues which I think will also fix your case per @alyssawilk

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto
Comment (noah8713) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:525]: Thanks for confirming. Will keep it only in v3

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (alyssawilk) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: +1 - if it could be in response_headers_to_add and be cleared for non-HTTP/1.1 that'd be fantastic.

Comment (noah8713) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: > I think this needs to be a much longer explanation. Specifically, it should clear say which RFC this violates, why it violates it, and when someone would want to set this.

Thanks yes I had it in commit details and should have extended in comments too. Will do part of amendments

Comment (noah8713) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: > +1 - if it could be in response_headers_to_add and be cleared for non-HTTP/1.1 that'd be fantastic.

Thanks @mattklein123 @alyssawilk seems this seems to be good path forward and looks neat. Saves us api change. So I will amend the logic to validate 2 things

1. if protocol h11 and request_headers has connection:keep-alive, retain connection-keep-alive
2. if protocol h11 and response_headers has keep-alive: timeout=xx, retain keep-alive response header. 


Will rectify the change logs accordingly part of amendments.

Comment (alyssawilk) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: I don't think that's what we're suggesting.   Always doing pass-through would be a very negative change for most Envoy operatators.
Envoy has a mechanism to add headers: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto request_headers_to_add / response_headers_to_add.
I'm suggesting that if what you need is to send a keep-alive header, _and_ if configuring it on a per-route or a per-hcm basis is sufficient, we should use that existing mechanism (iff the headers are being sent over an HTTP 1.1 codec)

If you absolutely need to preserve a timeout that Envoy will not actually enforce, you will need a different mechanism.  In that case I'd suggest  general 'do not clear hop by hop headers' config option, and you can use a filter to clear the ones you want.  I really would like to avoid having a knob for standards-breaking preservation of individual hop by hop headers.

Comment (noah8713) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: Thanks.  I dont think its  a new knob. I was referring to existing request_headers as request_headers_to_add which is already exposed (my bad should have explicitly stated the actual vars request_headers_to_add and response_headers_to_add ). Adding any custom headers are all good as long as envoy is not discarding in code as is unlike for connection and keep-alive. Just leveraging request_headers_to_add / response_headers_to_add as istio too provides changing headers via [virtual services](https://istio.io/latest/docs/reference/config/networking/virtual-service/#Headers) which underlying uses https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto  or we can use it for entire mesh with envoyfilter to apply these settings for all (thats upto users and operators to control). What I think is we still need to ensure its present in headers if users dont explicitly set it. However,  that too will not work in current case as its being discarded in code. So  will just enhance the logic to see if response_header_to_add has keep-alive and its h11 protocol keep it; Discard for other protocols. Also `Connection:keep-alive` is for request not for response and clear hop by hop only clears connection header in response not request. .Because its not just clear hop-by-hop knob, also  https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L102  and https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L93. needs to be amended.  Let me know if I misunderstood from code side as also recall you have already called it out https://github.com/envoyproxy/envoy/issues/17613#issuecomment-896103616 need to filter between the ones set by user with request/response_headers_to_add vs from codec. Its just response_header_to_add is per listener vs global change. That is why adding via config knob or runtime guard was also valid one to avoid mass refactor. Let me know else I am ok to keep the HCM config too to avoid over complicating/over engineering

Comment (noah8713) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: > If setting specific connection headers would be sufficient for this user's use case I'd much rather go that route. @noah8713 would set keep-alive work for you or do you actually need preservation of different headers from different upstreams?

Other custom headers are ok. Just need to retain these two connection:keep-alive in request and keep-alive in response as its hardcoded in code without protocol check or any knob control.


e.g. 
```
"response_headers_to_add": [
         {
          "header": {
           "key": "keep-alive",
           "value": "timeout=60"
          }
         },
         {
          "header": {
           "key": "x-test-me",
           "value": "ok-fine"
          }
         }
```

but because its hard coded in code, it gets discarded 
```
 curl --http1.1 -s -I localhost:10000
HTTP/1.1 200 OK
xxxx
x-test-me: ok-fine
```

Main reason is to make it easier for wider community to leverage the same and also saves us to maintain it in house for each envoy upgrade. That is why kept it per HCM config or runtime guard. Using HCM was beneficial to have it configure per listener. Amending via response_header_to_add field will need to iterate router config and it seems for one header need to do too much refactoring :) . Anyways I am ok to sync up during next meetup too and discuss but this is what we have in hand with all option. Please confirm accordingly @mattklein123 @alyssawilk :)

Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: I think what @alyssawilk is proposing which I agree with is to make request/response headers to add respect the connection headers and not remove them. If this works for your use case I think that is optimal.

Comment (noah8713) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:895]: Thanks @mattklein123 I did spend some time reading and tryout few bits but I am not clear from code at-least as clear_hop_by_hop removes it anyways ref https://github.com/envoyproxy/envoy/pull/17110 its only allowed for envoy mobile to passthrough list of these headers; verified multiple times it doesn't work as is even on setting it on req/response headers to add.  Our goal here is same to  passthrough the response `keep-alive:timeout=xx` settings from server/upstream as is to downstream so that connection can be reused as different apps have different keep-alive:timeout settings. Since clear_hop_by_hop is not exposed for HCM to turn it off or even over-ride;  per suggestion from @alyssawilk ask is to not change hop by hop logic for this case.  That is why exposed this change via config knob to return as is via first cut. Can we keep the config knob via hcm for time being? Later on we can amend setting for each application behind proxy via response_header_to_add and amend the clear hop by hop logic if set. Basically upstream/server can set that part of onboarding their service behind proxy and envoy can keep honoring it for http1.1. Not sure if it can be done without any major code changes but open to suggestions to move forward.

**Comment by noah8713** [comment]
Summary:
ref https://github.com/envoyproxy/envoy/pull/30621#discussion_r1441141319 ; trying request_header_to_add and response_header_to_add doesnt work because clear_hop_by_hop PR https://github.com/envoyproxy/envoy/pull/17110 removes it anyways. So ask is to make this thing work, can I amend the clear_hop_by_hop logic to not discard keep-alive if its present in response_header_to_add (just for http1.1) cc @mattklein123

**Comment by mattklein123** [comment]
> ref https://github.com/envoyproxy/envoy/pull/30621#discussion_r1441141319 ; trying request_header_to_add and response_header_to_add doesnt work because clear_hop_by_hop PR https://github.com/envoyproxy/envoy/pull/17110 removes it anyways. So ask is to make this thing work, can I amend the clear_hop_by_hop logic to not discard keep-alive if its present in response_header_to_add (just for http1.1) cc @mattklein123

Yes I believe we agreed that if the connection headers are explicitly added in response_headers_to_add we will not remove them in the various functions that clean them. @alyssawilk is this correct?

**Comment by SpecialYang** [comment]
Can we pick up this pr?

**Comment by SpecialYang** [comment]
Can we pick up this pr?

**Pull request closed without merging**
